Assets/Scripts/Models/HealthSystem.cs

```diff
@@ -0,0 +1,117 @@
+﻿using System;
+
+[Serializable]
+public class HealthSystem
+{
+    //all of this data should probably be generated at runtime or loaded from a file
+    public float Value
+    {
+        get
+        {
+            return _currentHealth;
+        }
+
+        private set
+        {
+            if (Invincible == true) return;
+            //We have been Revived!
+            if (_currentHealth <= 0 && value > _currentHealth)
+            {
+                _currentHealth = value;
+
+                if (_OnRevive != null)
+                    _OnRevive();
+                return;
+            }
+            //we have been hit but we are not dead!
+            if (value < _currentHealth && value > 0)
+            {
+                _currentHealth = value;
+                if (_OnHit != null)
+                    _OnHit();
+            }
+            //we have been healed!
+            if (value > _currentHealth)
+            {
+                if (value > MaxHealth)
+                    value = MaxHealth;
+                _currentHealth = value;
+                if (_OnHeal != null)
+                    _OnHeal();
+
+            }
+            //We have died and this means we have been hit!
+            if (value <= 0 && IsAlive)
+            {
+                _currentHealth = 0;
+                if (_OnHit != null)
+                    _OnHit();
+                if (_OnDestruction != null)
+                    _OnDestruction();
+            }
+        }
+    }
+    public float MaxHealth { get; set; }
+    public bool Invincible { get; set; }
+    public bool Healable { get; set; }
+    public bool IsAlive { get { return Value > 0; } }
+
+    private float _currentHealth = 100.0f;
+    private Action _OnHit;
+    private Action _OnHeal;
+    private Action _OnRevive;
+    private Action _OnDestruction;
+
+    public HealthSystem(float currentHealth, float maxHealth, bool invincible, bool healAble)
+    {
+        _currentHealth = currentHealth;
+        MaxHealth = maxHealth;
+        Invincible = invincible;
+        Healable = healAble;
+    }
+
+    public void RegisterOnHit(Action act)
+    {
+        _OnHit += act;
+    }
+    public void RegisterOnHeal(Action act)
+    {
+        _OnHeal += act;
+    }
+    public void RegisterOnRevive(Action act)
+    {
+        _OnRevive += act;
+    }
+    public void RegisterOnDestruction(Action act)
+    {
+        _OnDestruction += act;
+    }
+
+    public void UnregisterOnHit(Action act)
+    {
+        _OnHit -= act;
+    }
+    public void UnregisterOnHeal(Action act)
+    {
+        _OnHeal -= act;
+    }
+    public void UnregisterOnRevive(Action act)
+    {
+        _OnRevive -= act;
+    }
+    public void UnregisterOnDestruction(Action act)
+    {
+        _OnDestruction -= act;
+    }
+
+    public static HealthSystem operator+(HealthSystem h1, float num)
```

Comment:
`O[u]` is called an indexer. It has a get and set. It’s exactly the same as a property, so it can indeed change the object.

 

`i++` is shorthand for `i += 1`, which in turn is shorthand for `i = i + 1;` So *strictly speaking* it does not change the object it creates a new object and assigns it to the variable name from before. (I am speaking on a language level. In practice it often changes the object for performance reasons)

 

From: Alexander Fast [mailto:notifications@github.com] 
Sent: Saturday, August 20, 2016 3:07 PM
To: TeamPorcupine/ProjectPorcupine <ProjectPorcupine@noreply.github.com>
Cc: Oswald Maskens <ozie@live.be>; Mention <mention@noreply.github.com>
Subject: Re: [TeamPorcupine/ProjectPorcupine] Health system (#363)

 

In Assets/Scripts/Models/HealthSystem.cs <https://github.com/TeamPorcupine/ProjectPorcupine/pull/363#discussion_r75583246> :

> +        _OnHit -= act;
> +    }
> +    public void UnregisterOnHeal(Action act)
> +    {
> +        _OnHeal -= act;
> +    }
> +    public void UnregisterOnRevive(Action act)
> +    {
> +        _OnRevive -= act;
> +    }
> +    public void UnregisterOnDestruction(Action act)
> +    {
> +        _OnDestruction -= act;
> +    }
> +
> +    public static HealthSystem operator+(HealthSystem h1, float num)

@OzieGamma <https://github.com/OzieGamma>  your specific example is absolutely correct, But some operators do change the object they operate on. Like i++ or o[u] = 0, I realize now that Ive forgot the names for these. :S

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/TeamPorcupine/ProjectPorcupine/pull/363/files/26885cc752080e897219b1bc2d7a1bfe0a098054#r75583246> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AH2O3ZiQELJoOr7UI1oLNKlLXumjwTOhks5qh1BKgaJpZM4JpCRY> .  <https://github.com/notifications/beacon/AH2O3YxNpObtiryOOhXk16wwtW3MUhZrks5qh1BKgaJpZM4JpCRY.gif>